### ISSUE TITLE
`Tee-Object` should have a `-Console` parameter

### ISSUE THREAD

**Issue by iRon7** [PowerShell/PowerShell#19827]
### Summary of the new feature / enhancement

From the [**PowerShell for every system!**](https://github.com/PowerShell/PowerShell) idea, `Tee-Object` shouldn't need a system specific value to output to a console (and the pipeline) but simply tee the object to both **success stream** and **information stream**.

### Proposed technical implementation details (optional)

Wishful thinking:
(see: [Can you redirect Tee-Object to standard out?](https://stackoverflow.com/a/71040823/1701026))

```PowerShell
Get-Content data.txt | Tee-Object -Console | data_processor.exe
```

**Comment by ThomasNieto** [comment]
The parameter should be called `Host` since there could be a custom PS host that isn't a console and it also matches `Out-Host` writing to the `Information` stream.

**Comment by SeeminglyScience** [comment]
> The parameter should be called `Host` since there could be a custom PS host that isn't a console and also matches `Out-Host` writing to the `Information` stream.

It doesn't write to the `Information` stream believe it or not. Agree about the parameter name though (if the commands WG accepts the suggestion)

**Comment by ThomasNieto** [comment]
> It doesn't write to the Information stream believe it or not. Agree about the parameter name though (if the commands WG accepts the suggestion)

That is interesting didn't know that. I meant to put `Write-Host` instead of `Out-Host` in my comment.

**Comment by tolgabalci** [comment]
Earlier today I googled for a way to output to console with Tee-Object and this is what Google's predictive ai (which comes up on top of normal search results when enabled) said I can do:

![image](https://github.com/PowerShell/PowerShell/assets/2467618/c5485ebd-7aec-48ca-8fc8-7e74448bbe61)

I was super excited about using it, but sadly it doesn't work and very interestingly I can't find any references anywhere to discussions for adding a "-OutputConsole" flag.  It is almost like Google's predictive ai is anticipating its creation in the future.

I do think -OutHost would be a better name, and I do agree this is a much needed feature.  The documentation for Tee-Object actually provides an example on how to do this (documentation calls it "output to console"):
https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/tee-object?view=powershell-7.3#example-4-print-output-to-console-and-use-in-the-pipeline

However this method of outputting to the host is not consistent with other means of outputting in PowerShell.

I volunteer an additional common use case where this is useful:

Many times before outputting your results to their final pipeline destination (such as Out-File) you want to do some post-processing such as Sort-Object:
`1,4,21,3,9 | % { sleep 3; $_ } | Sort-Object | Tee-Object MyNumbers.txt`

The above works as desired, and we still output to the host as well for better debugging/operational awareness.  However, we are left to wait until the entire process is finished before we see any results on the host.

Often I run commands against many elements that will take many minutes to complete.  Not knowing if my output looks right until all the commands run against all elements is not ideal.  To eleviate the issue I now do the following:

`1,4,21,3,9 | % { sleep 3; $_ } | Tee-Object ($IsWindows ? '\\.\CON' : '/dev/tty') | Sort-Object | Out-File MyNumbers.txt`

This lets me begin seeing the host output immediately, as the commands occur, and I still get my sorted list outputted to a file at the end.

However, as you can see, the current Tee-object command with output to host is terse and would be much better written and read with a -OutHost flag.

**Comment by TobiasPSP** [comment]
The cmdlet working group reviewed this and finds it a valuable addition.  We believe it's appropriate to call the parameter -Host because there are other hosts beside the console, i.e. VSCode.

**Comment by dkaszews** [comment]
@mklement0 who is also a contributor [posted about it on stackoverflow a year and a half ago](https://stackoverflow.com/questions/71039495/can-you-redirect-tee-object-to-standard-out/71040823#71040823), I think implementing it inside existing `Tee-Object` should be even easier.

**Comment by ArmaanMcleod** [comment]
@mklement0 @TobiasPSP @iRon7 @SeeminglyScience 

I think `-Host` is not the best name since it will try to hide inherited [`PSCmdlet.Host`](https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.pscmdlet.host?view=powershellsdk-7.3.0#system-management-automation-pscmdlet-host) property when defining the parameter in the class deriving `PSCmdlet`.

We would probably get this compiler warning [`Compiler Warning (level 2) CS0108`](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0108?f1url=%3FappId%3Droslyn%26k%3Dk(CS0108)).

Thoughts?

**Comment by iSazonov** [comment]
What if we make the requested behavior _default_? It is more user friendly for interactive console scenarios (less typing) and it is not breaking change.

**Comment by ArmaanMcleod** [comment]
> What if we make the requested behavior _default_? It is more user friendly for interactive console scenarios (less typing) and it is not breaking change.

@iSazonov that's probably a better option. Can't think of any ways making host the default would cause a breaking change.

**Comment by dkaszews** [comment]
> Can't think of any ways making host the default would cause a breaking change.

Only if someone wrote a script wanting the user to type in the file name every time, without ability to give it as parameter beforehand - not a usecase I would personally care about.

**Comment by mklement0** [comment]
@ArmaanMcleod, regarding the `Host` shadowing problem: I think using `new` to silence the warning and and `base.Host` to access the inherited property should be fine in this case. 

While using `Host` as a parameter _alias_ is an option, it would be the true parameter name that shows in the syntax diagram, which strikes me as undesired (and `ConsoleHost` is a bit awkward, since the general concept is _host_, and a _console_ is just _one_ type of host).

As for making `-Host` the default: 
It sounds like a good idea, but note that the `tee` Unix utility does _not_ act that way by default; without arguments, it simply acts like `cat`, i.e. is in effect a _no-op_.

**Comment by TobiasPSP** [comment]
Let’s take a sec and re-define what the desired use case for tee-object is before we go into implementation questions.
(A) historically, it was meant to split the *output* stream #1 to a variable (which cannot easily be done elsehow)
(B) the stream could also be written to a file (which again isn’t trivial to do from within the pipeline in realtime unless you touch SteppablePipeline)
(C)  right now, we discuss echoing stream #1 to the host, for example to provide status or progress information. This is **not** a unique feature. It can be added easily using basic pipeline mechanics, i.e. | foreach-object { write-host $_; $_ }, and you can turn it into a filter or function with process block if you want this functionality to be permanent/a distinct command.
So when this new feature request was posted, the cmdlet WG reviewed and found that - even though it really adds nothing new, and echoing to the host isn‘t necessarily tee-ing anymore - this may indeed be *convenient* to some users, and that -Host sounds like a consistent parameter naming to enable this functionality. So if anyone cared to implement it, the WG would support it.
I think you shouldn’t invest too much energy at this point, and definitely not touch default behavior. Make it more complete for Linux people to feel welcomed, yet also consider that most of what was discussed or desired can in fact be done natively with the object oriented and scriptblock-based PS pipeline system, and there is a reason why Tee-Object historically is the least-used built-in cmdlet of all.

**Comment by TobiasPSP** [comment]
> What if we make the requested behavior _default_? It is more user friendly for interactive console scenarios (less typing) and it is not breaking change.

If you use it in a script that splits pipeline data silently to a variable (as originally intended by the cmdlet design), the cmdlet currently would not output anything visibly. 

Do you suggest in this scenario the cmdlet should now by default also echo the data to the console? The same console that is also used by transcripts? Or did I misunderstand the intended concept?

**Comment by mklement0** [comment]
First things first: Leaving the default behavior alone is fine in my estimation; I'm just making general points below:

Re (A): At least for cmdlets / advanced scripts/function it can more easily and efficiently be done with `-OutVariable`, which I suspect is one of the reasons `Tee-Object` isn't used all that much.

Re (C): In general, I think the fact that something can be done with a _custom_ implementation, using existing features, is per se never a good argument against introducing a _convenience feature of general interest_ that is _built in_.

>  isn‘t necessarily tee-ing anymore 

It's still teeing - it tees pipeline input to the host, just like the `tee` utility does with `...  | tee /dev/tty` (e.g. ` 'a'..'c' |  tee /dev/tty | cat -n`). What the metaphor of _teeing_ (borrowed from "T" piping in pipe-fitting) is simply that a "side stream" is being created in addition to the normal pipeline flow - irrespective of what that side stream's target is.

> Do you suggest in this scenario the cmdlet should now by default also echo the data to the console? 

No, I think the suggestion is to use `-Host` _implicitly_ if _neither `-Variable` nor `-FilePath` are specified_.

In other words: ` 'a'..'c' | Tee-Object | cat -n` would then act similarly to  `'a'..'c' | tee /dev/tty | cat -n`

The `tee` utility's default behavior (if no tee target is specified) is pointless; as noted, it just an unnecessary no-op. 
Having `Tee-Object` default to `-Host` is therefore an improvement.

On the flip side, conceptually this default would amount to the `-Host` switch _defaulting to `$true`_, which is generally discouraged.

**Comment by tolgabalci** [comment]
Since I originally commented above about 3 months back, I've been thinking a lot about this use case and I'd like to ask here what people think.

I now think Tee-Object may be unnecessary and may confuse PowerShell users to approach problems outside of the constructs PowerShell already provides.

I see the issue [20123 ](https://github.com/PowerShell/PowerShell/issues/20123) already going down this path.

Enough talk, let's look at some examples.

First the Tee-Object -Host feature:

We can currently write:
`"Zero", "One", "Last" | Tee-Object ($IsWindows ? '\\.\CON' : '/dev/tty') | Sort-Object
`

Proposal (Name of parameter TBD) is:
`"Zero", "One", "Last" | Tee-Object -Host | Sort-Object
`

However, this is still easily written in PowerShell without the use of Tee-Object as:
 `"Zero", "One", "Last" | % { Write-Host $_; $_ } | Sort-Object
`

Which can be written also as:
`"Zero", "One", "Last" | % { Write-Host $_; Write-Output $_ } | Sort-Object
`

All of these have the same processing and output:
```
Zero
One
Last
Last
One
Zero
```

This seems a bit too terse at first, but it uses PowerShell's built-in constructs and hence more flexible.  For example, you can use the same similar process to implement what's being asked in [20123 ](https://github.com/PowerShell/PowerShell/issues/20123):

```
"First", "Item 2", "Last" | % -Process {
    ($_ | % { "Pipeline 1 says '$_'" });
    ($_ | Out-File -Append foo.txt);
    ($_ | % { "Pipeline 3 says '$_'" });
}

Output:
Pipeline 1 says 'First'
Pipeline 3 says 'First'
Pipeline 1 says 'Item 2'
Pipeline 3 says 'Item 2'
Pipeline 1 says 'Last'
Pipeline 3 says 'Last'

foo.txt
First
Item 2
Last
```
The above spawns out three separate pipelines.

You may notice above that the -Process is default and does not need to be included.  I included it because I think it's funny Issue 20123 is suggesting its addition to Tee-Object when you can already do the same with For-EachObject.

So I'm thinking maybe it would be best served not to try to add functionality to Tee-Object to replicate functionality PowerShell already has built-in concepts for.

I don't mind adding functionality to cmdlets to make use easier, however, I think it may have the adverse effect of causing users to never discover the built-in PowerShell concepts.

**Comment by TobiasPSP** [comment]
regarding -Host as a useful convenience function, we are on the same page, and Cmdlet WG approved.

regarding anything else, let’s always avoid widening the use case of cmdlets too much. Cmdlets should be narrow but deep: clearly defined distinct purpose (narrow), parameters handle all typical use cases for this particular narrow purpose (wide). No universal super cmdlets.

This way we avoid overlap and ambiguities.

Isn‘t feature request #20123 that you reference really describing the existing functionality found in Foreach-Object?

**Comment by mklement0** [comment]
Agreed, @TobiasPSP; therefore, @tolgabalci, to recap:

Per [the WG decision](https://github.com/PowerShell/PowerShell/issues/19827#issuecomment-1642367665):

* The parameter name `-Host` [has already been decided on]...

* ... and implementing it been green-lit. 

The subsequent discussion originally related only whether to _tweak_ its behavior by making it _the default_.

At this point it may be best to:

* Implement the feature as approved.

* Open a _new issue_ for making `-Host` the implied default, assuming someone feels strongly enough about it (I personally don't).

---

To avoid further distractions, I'm hiding my responses to the two most recent comments in a collapsible `<details></details>` section.

<details>

> may confuse PowerShell users to approach problems outside of the constructs PowerShell already provides.

`Tee-Object` _complements_ the existing constructs. 
No one should be tempted to use it for anything other than its _core mandate_: to process the input objects _twice_ (that's the "teeing" part): once as usual, by sending it down the pipeline, and again by sending it to a _different_ target.

Adding a `-Host` target - as a much more convenient  alternative to `Tee-Object ($IsWindows ? '\\.\CON': '/dev/tty')` and especially to `% { Write-Host $_; $_ }` (which is also slower) - is a _natural extension_ of the cmdlet's core mandate.

The same applies to adding a `-Process` parameter that allows _pre-processing_ the input before sending it to the tee target. It's yet another - powerful - convenience that fits in with and extends the _core functionality_ of `Tee-Object`.

To illustrate the problem with the anti-convenience stance (the question is rhetorical):
Do you think `Where-Object` should never have been implemented, given that anything you can do with it can also be done with `ForEach-Object`?

</details>

**Comment by TobiasPSP** [comment]
> Since I originally commented above about 3 months back, I've been thinking a lot about this use case and I'd like to ask here what people think.
> 
> I now think adding additional functionality to Tee-Object may be unnecessary and may confuse PowerShell users to approach problems outside of the constructs PowerShell already provides.
> 
> I see the issue [20123 ](https://github.com/PowerShell/PowerShell/issues/20123) already going down this path.
> 
> Enough talk, let's look at some examples.
> 
> First the Tee-Object -Host feature:
> 
> We can currently write: `"Zero", "One", "Last" | Tee-Object ($IsWindows ? '\\.\CON' : '/dev/tty') | Sort-Object `
> 
> Proposal (Name of parameter TBD) is: `"Zero", "One", "Last" | Tee-Object -Host | Sort-Object `
> 
> However, this is still easily written in PowerShell without the use of Tee-Object as: `"Zero", "One", "Last" | % { Write-Host $_; $_ } | Sort-Object `
> 
> Which can be written also as: `"Zero", "One", "Last" | % { Write-Host $_; Write-Output $_ } | Sort-Object `
> 
> All of these have the same processing and output:
> 
> ```
> Zero
> One
> Last
> Last
> One
> Zero
> ```
> 
> This seems a bit too terse at first, but it uses PowerShell's built-in constructs and hence more flexible. For example, you can use the same similar process to implement what's being asked in [20123 ](https://github.com/PowerShell/PowerShell/issues/20123):
> 
> ```
> "First", "Item 2", "Last" | % -Process {
>     ($_ | % { "Pipeline 1 says '$_'" });
>     ($_ | Out-File -Append foo.txt);
>     ($_ | % { "Pipeline 3 says '$_'" });
> }
> 
> Output:
> Pipeline 1 says 'First'
> Pipeline 3 says 'First'
> Pipeline 1 says 'Item 2'
> Pipeline 3 says 'Item 2'
> Pipeline 1 says 'Last'
> Pipeline 3 says 'Last'
> 
> foo.txt
> First
> Item 2
> Last
> ```
> 
> The above spawns out three separate pipelines.
> 
> You may notice above that the -Process is default and does not need to be included. I included it because I think it's funny that Issue 20123 is suggesting its addition to Tee-Object when you can already do the same with For-EachObject.
> 
> So I'm thinking maybe it would be best served not to try to add functionality to Tee-Object to replicate functionality PowerShell already has built-in concepts for.
> 
> I don't mind adding functionality to cmdlets to make use easier, however, I think it may have the adverse effect of causing users to never discover the built-in PowerShell concepts.

Thanks for pointing out your thoughts and your learning points over the past months.

This is right on. This is why the cmdlet WG is thoroughly chewing on requests - not to annoy or delay but in an effort to conserve the core nature of PowerShell while being open to new. 

With PS becoming increasingly available on other platforms, new ideas are posted. Some are valuable additions, including making features more discoverable.

Others are a result of applying „other shell“ muscle memory to PowerShell when in reality we need adoption to PS nature.

**Comment by TobiasPSP** [comment]
> 



> Agreed, @TobiasPSP; therefore, @tolgabalci, to recap:
> 
> Per [the WG decision](https://github.com/PowerShell/PowerShell/issues/19827#issuecomment-1642367665):
> 
> * The parameter name `-Host` [has already been decided on]...
> * ... and implementing it been green-lit.
> 
> The subsequent discussion originally related only whether to _tweak_ its behavior by making it _the default_.
> 
> At this point it may be best to:
> 
> * Implement the feature as approved.
> * Open a _new issue_ for making `-Host` the implied default, assuming someone feels strongly enough about it (I personally don't).
> 
> To avoid further distractions, I'm hiding my responses to the two most recent comments in a collapsible `<details></details>` section.
> 
> > may confuse PowerShell users to approach problems outside of the constructs PowerShell already provides.
> 
> `Tee-Object` _complements_ the existing constructs. No one should be tempted to use it for anything other than its _core mandate_: to process the input objects _twice_ (that's the "teeing" part): once as usual, by sending it down the pipeline, and again by sending it to a _different_ target.
> 
> Adding a `-Host` target - as a much more convenient alternative to `Tee-Object ($IsWindows ? '\\.\CON': '/dev/tty')` and especially to `% { Write-Host $_; $_ }` (which is also slower) - is a _natural extension_ of the cmdlet's core mandate.
> 
> The same applies to adding a `-Process` parameter that allows _pre-processing_ the input before sending it to the tee target. It's yet another - powerful - convenience that fits in with and extends the _core functionality_ of `Tee-Object`.
> 
> To illustrate the problem with the anti-convenience stance (the question is rhetorical): Do you think `Where-Object` should never have been implemented, given that anything you can do with it can also be done with `ForEach-Object`?

thank you for your careful thoughts and great summary.

**Comment by iRon7** [comment]
> However, this is still easily written in PowerShell without the use of Tee-Object as:
> `"Zero", "One", "Last" | % { Write-Host $_; $_ } | Sort-Object `

Aside from @mklement0 arguments. `Write-Host` is written for the end of the pipeline. There is a **Strongly Encouraged Development Guideline** that states that cmdlets should [implement for the Middle of a Pipeline](https://learn.microsoft.com/powershell/scripting/developer/cmdlet/strongly-encouraged-development-guidelines#implement-for-the-middle-of-a-pipeline). The reason for this is that cmdlets should also *be used* in the Middle of a Pipeline and with that [avoid wrapping cmdlet pipelines](https://learn.microsoft.com/powershell/scripting/dev-cross-plat/performance/script-authoring-considerations#avoid-wrapping-cmdlet-pipelines) for better performance.

I have written a small `TeeObject` prototype that wraps the original `Tee-Object` cmdlet:

<details>
  <summary>TeeObject</summary>

  ```powershell
function TeeObject {
    [CmdletBinding(DefaultParameterSetName='HostFormatted')]
    param(
        [Parameter(ValueFromPipeline=$true)]
        [psobject]
        ${InputObject},

        [Parameter(ParameterSetName='File', Mandatory=$true, Position=0)]
        [Alias('Path')]
        [string]
        ${FilePath},

        [Parameter(ParameterSetName='LiteralFile', Mandatory=$true)]
        [Alias('PSPath','LP')]
        [string]
        ${LiteralPath},

        [Parameter(ParameterSetName='File')]
        [switch]
        ${Append},

        [Parameter(ParameterSetName='File')]
        [Parameter(ParameterSetName='LiteralFile')]
        [ValidateNotNullOrEmpty()]
        [System.Text.Encoding]
        ${Encoding},

        [Parameter(ParameterSetName='Variable', Mandatory=$true)]
        [string]
        ${Variable},

        [Parameter(ParameterSetName='HostFormatted', Position=0)]
        [string]
        ${HostFormatted}
    )

    begin {
        $outBuffer = $null
        if ($PSBoundParameters.TryGetValue('OutBuffer', [ref]$outBuffer)) { $PSBoundParameters['OutBuffer'] = 1 }
        if ($PSBoundParameters.TryGetValue('HostFormatted', [ref]$outBuffer)) { $PSBoundParameters.Remove('HostFormatted') }
        if ($HostFormatted) {
            $Command = 'Microsoft.PowerShell.Utility\Write-Host'
        }
        else {
            if ($PsCmdlet.ParameterSetName -eq 'HostFormatted') { $PSBoundParameters['FilePath'] = $IsWindows ? '\\.\CON': '/dev/tty' }
            $Command = 'Microsoft.PowerShell.Utility\Tee-Object'
        }
        $wrappedCmd = $ExecutionContext.InvokeCommand.GetCommand($Command, [System.Management.Automation.CommandTypes]::Cmdlet)
        $scriptCmd = {& $wrappedCmd @PSBoundParameters }

        $steppablePipeline = $scriptCmd.GetSteppablePipeline($myInvocation.CommandOrigin)
        $steppablePipeline.Begin($PSCmdlet)
    }

    process {
        if ($HostFormatted) {
            $steppablePipeline.Process($ExecutionContext.InvokeCommand.ExpandString($HostFormatted))
            $_
        }
        else {
            $steppablePipeline.Process($_)
        }
    }

    end {
        $steppablePipeline.End()
    }

    clean {
        if ($null -ne $steppablePipeline) {
            $steppablePipeline.Clean()
        }
    }
}
  ```
</details>

This:
```PowerShell
Measure-Command {
    1..10000 | TeeObject | Out-Null
}
```
Is about `6` times faster than:
```PowerShell
Measure-Command {
    1..10000 | Foreach-Object { Write-Host $_; $_ } | Out-Null
}
```
<sup>(I suspect this is also due the fact that `Tee-Object ($IsWindows ? '\\.\CON': '/dev/tty')` is just faster than `Write-Host $_; $_` but just using the **steppablePipeline** on `Write-Host` is alsofaster)</sup>

Anyways, 

> While using `Host` as a parameter _alias_ is an option, it would be the true parameter name that shows in the syntax diagram, which strikes me as undesired (and `ConsoleHost` is a bit awkward, since the general concept is _host_, and a _console_ is just _one_ type of host).

<sup>(I bring this up as any **default** behavior might obstruct any other future implementations in this direction)</sup>

What if it is not a `switch` but a format `string`, e.g.: `-HostFormatted` (as implemented in the above prototype).
This will allow me to things along with:

```PowerShell
Get-ChildItem | TeeObject 'File: $_' | Select-String 'Host' | TeeObject 'Line: $($_.LineNumber)' | Set-Content .\Test.txt
```

**Comment by TobiasPSP** [comment]
thank you for sharing your thoughts.

Regarding your first part:
If you would like to stream output to the host in the middle of the pipeline, adding the discussed **-Host** parameter to *Tee-Object* makes sense and was already approved by the cmdlet WG. So that part we may want to put at rest. 

Regarding the actual implementation - switch parameter that echoes **-vs-** named parameter that allows for custom formatting:
my *personal* view (not speaking on behalf of the WG) is that the new feature should primarily simplify life. A switch parameter, plus maybe a default behavior when no parameter is specified, is short, convenient and self-explanatory. 

The performance differences you noticed are due to the different techniques used. If you replace | Foreach-Object { Write-Host $_; $_ } with | & {process{ Write-Host $_; $_ }} you‘d probably already see some improvement (simple functions perform much faster in a pipeline than advanced functions due to their much simpler pipeline binding).

If there is in fact a significant demand for being able to process the output prior to sending it to *-Host*, then my 5ct would be to try and implement it like this:
1. *Tee-Object* (without arguments) streams objects to host as-is
2. *Tee-Object -Host { $_ } (with scriptblock argument) streams objects and allows preprocessing, i.e. formatting, excluding results via *if* etc pp

**Comment by mklement0** [comment]
Note that  https://github.com/PowerShell/PowerShell/issues/20123 is the _generalization_ of `Tee-Object -Host { "File: $_" }`, via a _separate_ parameter, `-Process`, to receive the script block: 
* It allows you to use a script block to preprocess the inputs with _whatever target you choose_, `-Host`, `-Variable` or `-File` - e.g. `Tee-Object -Host -Process { "File: $_" }` or `Tee-Object -Variable outVar -Process { "File: $_" }` - or even with _no_ target, in which case success output, if any, would simply be discarded (e.g. `Tee-Object -Process { Export-Csv out.csv }` to export the inputs to a CSV file).
* For further convenience, the `[scriptblock]`-typed `-Process` parameter can be made positional, which would then directly enable `Tee-Object -Host { "File: $_" }` or `Tee-Object -Variable outVar { "File: $_" }`, for instance.
* A proof-of-concept is [here](https://github.com/PowerShell/PowerShell/issues/20123#issuecomment-1684236212), though there is an [unanswered technical question re potential loss of state when modifying the script block via _rebuilding_ it](https://github.com/PowerShell/PowerShell/issues/20123#issuecomment-1685308611).

Therefore, I suggest continuing the discussion at #20123

**Comment by iRon7** [comment]
@mklement0,
of cause, it indeed already comes all together in #20123 😊

**Comment by TobiasPSP** [comment]
Can I suggest we do the opposite? This request here was alteady approved by the WG and is ready to go. #20123 would probably have no merit because it is too close to Foreach-Object as a srand-alone feature request.Proceeding with this (approved) request and possibly incorporating the scriptblock extension „while at it“ seems a good path to go to me.Am 11.09.2023 um 14:59 schrieb iRon7 ***@***.***>:﻿
@mklement0,
of cause, it indeed already comes all together in #20123 😊

**Comment by iRon7** [comment]
@TobiasPSP,

(I am definitely not a linux user 😀)

> If you would like to stream output to the host in the middle of the pipeline in a performant way, adding the discussed -Host parameter to Tee-Object makes sense and was already approved by the cmdlet WG. So that part we may want to put at rest.

You waved this performance argument away but this is probably one biggest reasons you would use `Tee-Object -Process`. There are a lot of linked issues to #20123, including #20133 where it shows e.g. that there is no common `PassThru` parameter (aka implementation). Therefore if you would like to implement an end-of-pipeline cmdlet in the middle of the pipeline (or combine this with any other end-of-pipeline cmdlet in a single pipeline), you would like to be able to do this:

```PowerShell
... | Tee-Object -Process { Export-Csv .\Output.csv } | ...
```

As it is a [few hundred times faster](https://learn.microsoft.com/powershell/scripting/dev-cross-plat/performance/script-authoring-considerations#avoid-wrapping-cmdlet-pipelines) than any current obvious nested pipeline implementation

**Comment by TobiasPSP** [comment]
Got it. So you feel the topic here is just a use case of your more general idea and could be implemented by -Process { Out-Host }?

I can take #20123 to one of the upcoming WG meetings for discussion. To be frank, I cannot yet see **striking** practical use. Typically, if you really wanted to stream data to i.e. Export-Excel **in the middle of a pipeline**, you‘d probably use two pipelines. True, this would block streaming but what would be a practical case where this really matters? 

I am asking because I know how discussions go, and I have seen many *technically inspiring* ideas fall out of the roaster when there was only little mainstream use. So I‘d like as many good practical arguments as I could get to support your suggestion.

Echoing to the console in contrast - as discussed here and as already approved - provides many immediately visible use cases - i.e. for logging, debugging or just to check data in real time - that apply to general audience, so no „edge case“ argument that needs to be proven false.

**Comment by TobiasPSP** [comment]
I have tagged #20123 and added it to our Sep 20 meeting and hope we‘ll get to it. 
Meanwhile you may want to add **practical** convincing use cases plus a suggestion how the ideas we discussed in *this* thread could be merged.

**Comment by iRon7** [comment]
As described in #20693, it is not clear to me whether a issue that is in <kbd>Up-for-Grabs</kbd> state should be closed (before it is actually implemented).

**Comment by iRon7** [comment]
It is unclear to me whether an issue already in <kbd>up-for-graps</kbd> state shoud be allowed to (automatically) closed in the process of waiting for actual implementation...
(Hence this comment.)
